File: typescript/test/historical-selector-bounds.test.ts
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { afterEach, describe, expect, it, vi } from "vitest";

import { main } from "../src/cli.js";
import { TypeError as FGraphTypeError } from "../src/errors.js";
import { connect } from "../src/store.js";
import { INT64_MAX, INT64_MIN } from "../src/values.js";

const directories: string[] = [];

async function invoke(...args: string[]): Promise<{
  code: number;
  stdout: string;
  stderr: string;
}> {
  let stdout = "";
  let stderr = "";
  const stdoutSpy = vi.spyOn(process.stdout, "write").mockImplementation(((
    chunk: string | Uint8Array,
  ) => {
    stdout += chunk.toString();
    return true;
  }) as typeof process.stdout.write);
  const stderrSpy = vi.spyOn(process.stderr, "write").mockImplementation(((
    chunk: string | Uint8Array,
  ) => {
    stderr += chunk.toString();
    return true;
  }) as typeof process.stderr.write);
  try {
    return { code: await main(args), stdout, stderr };
  } finally {
    stdoutSpy.mockRestore();
    stderrSpy.mockRestore();
  }
}

afterEach(() => {
  while (directories.length > 0)
    rmSync(directories.pop() as string, { recursive: true, force: true });
  delete process.env.FGRAPH_CLOCK;
});

describe("historical selector integer bounds", () => {
  it("resolves an integer above transaction ids as an instant", () => {
    using db = connect(":memory:", { clock: 1_767_225_600_000_000n });
    const report = db.transact({
      id: "selector/subject",
      "item/value": "present",
    });

    expect(BigInt(report.at as number)).toBeGreaterThan(
      BigInt(report.tx as number),
    );
    expect(db.at(report.at).entity("selector/subject")["item/value"]).toBe(
      "present",
    );
  });

  it.each([INT64_MIN, INT64_MAX, INT64_MIN - 1n, INT64_MAX + 1n])(
    "rejects library selector %s with the typed taxonomy",
    (selector) => {
      using db = connect(":memory:", { clock: 1_767_225_600_000_000n });

      expect(() => db.at(selector)).toThrowError(FGraphTypeError);
    },
  );

  it("rejects CLI boundaries and adjacent overflows as typed errors", async () => {
    const directory = mkdtempSync(join(tmpdir(), "fgraph-selector-"));
    directories.push(directory);
    const path = join(directory, "selector.db");
    process.env.FGRAPH_CLOCK = "1767225600000000";
    expect(
      (
        await invoke(
          "add",
          '{"id":"selector/subject","item/value":"present"}',
          "--db",
          path,
        )
      ).code,
    ).toBe(0);
    const query = '{"find":["?e"],"where":[["?e","item/value","present"]]}';
    const selectors = [
      INT64_MIN.toString(),
      INT64_MAX.toString(),
      (INT64_MIN - 1n).toString(),
      (INT64_MAX + 1n).toString(),
    ];

    for (const command of [
      ["get", "selector/subject"],
      ["q", query],
    ]) {
      for (const selector of selectors) {
        expect(
          await invoke(...command, "--at", selector, "--db", path),
        ).toMatchObject({
          code: 1,
          stderr: expect.stringContaining("TypeError:"),
        });
      }
    }
  });
});

File: go/historical_selector_bounds_test.go
package fgraph

import (
	"context"
	"encoding/json"
	"errors"
	"strconv"
	"testing"
)

const (
	selectorInt64Min = -1 << 63
	selectorInt64Max = 1<<63 - 1
)

func TestHistoricalSelectorBounds(t *testing.T) {
	ctx := context.Background()
	db := fixedDB(t, ":memory:")
	report, err := db.Transact(ctx, E{"id": "selector/subject", "item/value": "present"})
	if err != nil {
		t.Fatal(err)
	}
	if report.At <= report.Tx {
		t.Fatalf("test needs an instant above the latest transaction id: %#v", report)
	}
	view, err := db.At(ctx, report.At)
	if err != nil {
		t.Fatal(err)
	}
	if entity, entityErr := view.Entity(ctx, "selector/subject"); entityErr != nil || entity["item/value"] != "present" {
		t.Fatalf("integer instant view = %#v, %v", entity, entityErr)
	}

	for _, selector := range []any{
		int64(selectorInt64Min),
		int64(selectorInt64Max),
		json.Number("-9223372036854775809"),
		json.Number("9223372036854775808"),
	} {
		if _, selectorErr := db.At(ctx, selector); !errors.Is(selectorErr, ErrType) {
			t.Errorf("At(%v) error = %v, want TypeError", selector, selectorErr)
		}
	}
}

func TestCLIHistoricalSelectorBounds(t *testing.T) {
	t.Setenv("FGRAPH_CLOCK", "1767225600000000")
	path := t.TempDir() + "/selector.db"
	if _, err := runCLIForTest(t, "", "add", `{"id":"selector/subject","item/value":"present"}`, "--db", path); err != nil {
		t.Fatal(err)
	}
	query := `{"find":["?e"],"where":[["?e","item/value","present"]]}`
	selectors := []string{
		strconv.FormatInt(selectorInt64Min, 10),
		strconv.FormatInt(selectorInt64Max, 10),
		"-9223372036854775809",
		"9223372036854775808",
	}
	for _, command := range [][]string{{"get", "selector/subject"}, {"q", query}} {
		for _, selector := range selectors {
			args := append(append([]string{}, command...), "--at", selector, "--db", path)
			if _, err := runCLIForTest(t, "", args...); !errors.Is(err, ErrType) {
				t.Errorf("%v error = %v, want TypeError", args, err)
			}
		}
	}
}
